confidant/scripts/archive.py

```diff
@@ -0,0 +1,169 @@
+        ),
+        Option(
+            '--ids',
+            dest='ids',
+            help=('Archive a comma separated list of credential IDs. (mutually'
+                  ' exclusive with --days)'),
+        ),
+    ]
+
+    def credential_in_service(self, _id, services):
+        for service in services:
+            if _id in service.credentials:
+                return True
+        return False
+
+    def save(self, saves, force=False):
+        _saves = ', '.join([save.id for save in saves])
+        if not force:
+            logger.info(
+                'Would have archived credential and revisions: {}'.format(
+                    _saves,
+                )
+            )
+            return
+        logger.info(
+            'Archiving credential and revisions: {}'.format(
+                _saves,
+            )
+        )
+        with CredentialArchive.batch_write() as batch:
+            for save in saves:
+                batch.save(save)
+
+    def delete(self, deletes, force=False):
+        _deletes = ', '.join([delete.id for delete in deletes])
+        if not force:
+            logger.info(
+                'Would have deleted credential and revisions: {}'.format(
+                    _deletes,
+                )
+            )
+            return
+        logger.info(
+            'Deleting credential and revisions: {}'.format(
+                _deletes,
+            )
+        )
+        with Credential.batch_write() as batch:
+            for delete in deletes:
+                batch.delete(delete)
+
+    def archive(self, credentials, force):
+        services = [
+            service for service in Service.data_type_date_index.query('service')
+        ]
+        for credential in credentials:
+            if self.credential_in_service(credential.id, services):
+                msg = ('Skipping archival of disabled credential {}, as it'
+                       ' is still mapped to a service.')
+                logger.warning(msg.format(credential.id))
+                continue
+            saves = []
+            deletes = []
+            # save the current record.
+            archive_credential = CredentialArchive.from_credential(
+                credential,
+            )
+            saves.append(archive_credential)
+            # fetch and save every revision
+            _range = range(1, credential.revision + 1)
+            ids = []
+            for i in _range:
+                ids.append("{0}-{1}".format(credential.id, i))
+            revisions = Credential.batch_get(ids)
+            for revision in revisions:
+                archive_revision = CredentialArchive.from_credential(
+                    revision,
+                )
+                saves.append(archive_revision)
+                deletes.append(revision)
+            deletes.append(credential)
+            try:
+                self.save(saves, force=force)
+            except Exception:
+                logger.exception(
```

Comment:
If we're productionizing this we should add a stat here to alarm on

Reply:
Ah. Was assuming this would be run by hand, but I guess I didn't consider that we could possible run this in a cron... I'll add a stat.